Dashboard Cards: Display 3 latest pages card

AjeshRPai opened the pull request:
## Related Issue  
Closes #18212 

## Description
This PR adds support for showing the Pages item in pages card. 

## Explanation of Changes 
- ⊕ Adds the UI for pages item in Pages card in Light and Dark mode 
- ⊕ The logic for showing the status icon and text according to the page status 
- ⊕ The logic for showing Last modified or Scheduled time in pages card according to status 
- ⊕ Adds unit tests in [PagesCardBuilderTest.kt](https://github.com/wordpress-mobile/WordPress-Android/pull/18337/files#diff-a78feba5d939a847e3b38b5c662e34a2fef3846f10613e118bf60e80a6983681)  
- ⊘ Does not implement the click listener for the page items - will be taken up in a separate PR
- ⊘ Does not implement any tracking logic 

## 📷  Screenshots

### Pages card 
|Light Mode | Dark Mode| 
|---|---|
|  ![screenshot_pages_card_lightMode_](https://user-images.githubusercontent.com/17463767/235445580-0879536b-40ad-4b0e-8f62-ade2757a0ab4.png) | ![screenshot_pages_card_night_mode_](https://user-images.githubusercontent.com/17463767/235445598-ba1ad312-cef8-43d3-8b72-a844fbaee9ec.png) |  

### Scheduled relative time 
👁️ Note the relative time in Scheduled Page item in Card

| In Minutes  | In Hours |  In Days | In more than a week |
|---|---|---|---|
|![screenshot_scheduled_in_minutes_](https://user-images.githubusercontent.com/17463767/235447760-68a13380-a1db-47f3-9ee6-12e6c4362e5a.png)|![screenshot_scheduled_in_hours_](https://user-images.githubusercontent.com/17463767/235445716-70b37b0f-531c-4e9b-937c-145e7330139f.png)|![screenshot_scheduled_in_days_](https://user-images.githubusercontent.com/17463767/235445700-5a35ae28-1f37-4145-9a4b-1eaafe1a0371.png)|![screenshot_scheduled_in_months_](https://user-images.githubusercontent.com/17463767/235445734-7ca10ba2-ef0f-4e39-83a7-7ab4e0eb8fdf.png)| 

## Test 

#### Prerequisite
<details><summary>◐ Toggle the <code>dashboard_card_pages</code> flag</summary>

1. Go to `Me` → `App Settings` → `Debug settings`
4. Scroll to the `Features in development` section
5. Tap on the item corresponding to the flag
6. Tap `RESTART THE APP` button
<img width="501" alt="Screenshot 2023-04-27 at 10 59 09 AM" src="https://user-images.githubusercontent.com/17463767/234767755-65dfe2fa-c6d0-475c-b775-67004aa65ebd.png">

</details>

### Card UI is as expected in the Project spec and Designs
Project spec - pe7hp4-gi-p2
Design p2 - pcdRpT-2bP-p2
Design ref Figma - YSMw2nbFAgPpjZLLbYvTfT-fi-310_3783

- Go to app → Dashboard 
- ✅ Verify that the Pages are shown with the correct title 
   - Long titles are truncated 
   - If no title is present for a page, then **Untitiled** is shown
-  ✅  Verify that the correct icon and text are shown for the Page status
   - Published, Draft and Scheduled
- ✅ Verify that the correct time is shown next to the status icon and text 
   - For **Published and Draft page** →  the last modified time is shown 
       - Modified date is Less than an hour → N minutes ago 
       - Modified date is Less than a day but more than an hour → N Hours ago 
       - Modified date is Less than a week →  N Days ago 
       - Modified date is More than 1 week → Date is shown
   - For **Scheduled Page** → the relative Scheduled time is shown
       - Scheduled date is in Less than an hour → In N minutes
       - Scheduled date is Less than a day but more than an hour → In N hours
       - Scheduled date is Less than a week →  In N Days
       - Scheduled date is More than 1 week → Date is shown

💡 Note: To test the relative time, modify or schedule a page and go to Settings on device and change the device time.
 

### Verify the Unit tests 
Code review the unit tests in [PagesCardBuilderTest.kt](https://github.com/wordpress-mobile/WordPress-Android/pull/18337/files#diff-a78feba5d939a847e3b38b5c662e34a2fef3846f10613e118bf60e80a6983681) pass, sufficient and doesn't miss any scenario

## Regression Notes
1. Potential unintended areas of impact
Pages card is not shown as expected 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and Unit tests 

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

AjeshRPai committed:
Adds: style for pages item title in pages card

AjeshRPai committed:
+ Adds:the logic to get the status text for pages item

+ Adds: the logic to get the status text
+ Adds: the string resources for status text in pages card

AjeshRPai committed:
+ Adds:the logic to get the status icon for pages item

+ Adds: the logic to get the status icon
+ Adds: the drawable resources for status icon in pages card
* Refactors: the status icon to be nullable in case status is
not identified

AjeshRPai committed:
Fixes: the string value of PagesCardContentType

AjeshRPai committed:
* Updates: the status mapping card builder

Updates: the status icon and text mapping of the pages card item with
 correct PagesCardContentType.kt

AjeshRPai committed:
Adds: test for published status check

AjeshRPai committed:
+ Adds: test for page status - draft and scheduled

AjeshRPai committed:
* Updates: the logic of building the create Page card

Updates: the logic of building the create page card with
mapped page items instead of api response
Updates: the params of create page card with footer click function
instead of the entire params

AjeshRPai committed:
Updates: the logic of mapping the pages items

Updates: the logic of creating the pages items from api response by
filtering the supported status in pages card

AjeshRPai committed:
[WIP] Adds: logic for setting the status icon and text

AjeshRPai committed:
Updates: function statement for readability

AjeshRPai committed:
Updates: text style for status and updates color of published icon

AjeshRPai committed:
Adds: text view for last edited time or scheduled time

AjeshRPai committed:
Adds: long title in preview to check how long title behave

AjeshRPai committed:
[WIP] + Adds: the logic for showing the last edited or scheduled time

Adds: the logic for getting the last edited or scheduled time using
date or modified time depending on status
Adds: the logic in PagesItemViewHolder to show the data

AjeshRPai committed:
+ Adds: Missing dependency for builder in test class

* Makes test statement shorter to fix lint errors
+ Adds mock for date time utils wrapper dependancy

AjeshRPai committed:
* Updates: getLastEditedOrScheduledTime fun to make it readable

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/pages/PagesCardBuilder.kt:
```diff
@@ -49,59 +56,68 @@ class PagesCardBuilder @Inject constructor(
-    private fun createNewPageCardWithAddPageMessage(params: PagesCardBuilderParams): CreatNewPageItem {
+    private fun createNewPageCardWithAddPageMessage(onFooterLinkClick: () -> Unit): CreatNewPageItem {
         return CreatNewPageItem(
             label = UiString.UiStringRes(R.string.dashboard_pages_card_no_pages_create_page_button),
```
(early np - since this is a draft) but you could statically import the namespace and drop the `UiString` prefix.

AjeshRPai replied:
Nice, Thanks. Done in 574b12e23c8a85386ec13c6966194ea71e5139b6

zwarm commented on WordPress/src/main/res/layout/pages_item.xml:
```diff
@@ -17,8 +17,37 @@
+        tools:text="@string/dashboard_card_page_item_status_published"
+        android:drawablePadding="@dimen/margin_medium"
+        app:layout_constraintStart_toStartOf="parent"
+        android:paddingStart="10dp"
```
np (again, a bit early) - just a reminder to use `android:paddingHorizontal` and `android:paddingVertical`. :)

AjeshRPai replied:
Yup, Thanks for point it out. Done in [b4269b0](https://github.com/wordpress-mobile/WordPress-Android/pull/18337/commits/b4269b060abe8dc42ea4c0e36c0881d20840d884)

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/pages/PagesCardBuilder.kt:
```diff
@@ -49,59 +56,68 @@ class PagesCardBuilder @Inject constructor(
     }
 
-    private fun getCreatePageCard(params: PagesCardBuilderParams): CreatNewPageItem {
+    private fun getCreatePageCard(pages: List<PageContentItem>, onFooterLinkClick: () -> Unit): CreatNewPageItem {
```
Nice refactor ❤️

AjeshRPai committed:
+ Adds: UI string import statically for readability

AjeshRPai committed:
+ Adds: the logic to show scheduled date in relative time span

+ Adds: the logic to return relative time span for future dates in
 DateTimeUtilsWrapper.kt

AjeshRPai committed:
- Removes: Unneeded UiString from the statement and import

AjeshRPai committed:
+ Adds: UI string import statically for readability

- Removes: redundant UiString in all the usages

AjeshRPai committed:
+ Adds: test for relative time shown on card

+ Adds: test for scheduled and published page relative time shown

AjeshRPai committed:
* Updates: statements and models used in test class

* Updates: test statements by moving the common mock to setUpMocks
function
* Updates: Models used in tests with status identifier

AjeshRPai committed:
+ Adds: PagesCardBuilderParams import statically for readability

- Removes: redundant MySiteCardAndItemBuilderParams in all the usages

AjeshRPai committed:
+ Adds: Helper for PagesCardBuilderParams function to test class

- Removes: PagesCardBuilderParams in tests with helper function
* Updates: the test statement to invoke BuilderParams generator
function to only build with relevant model(Scheduled/Published/Drafts)
 instead of all the models

AjeshRPai committed:
*[WIP] Updates: the xml for Pages item in pages card

+ Adds: the list divider element to the xml for pages item in card
* Updates: the padding attributes and margin attributes with
paddingHorizontal, paddingVertical and marginHorizontal

AjeshRPai committed:
* Updates: the naming of icon resources used to show status of page

AjeshRPai committed:
* Implements: the background for page status text view in light/dark

AjeshRPai committed:
* Implements: the colors for pages card components in light/dark

* Implements: the light/dark text color for status and
 last modified text
* Updates: the status icon color on light and dark mode

AjeshRPai committed:
^ Replaces: hardcoded dimensions in pages_item with values from dimens

Adds: dimension values for pages item components
Replaces: the hardcoded references with values from dimens.xml

AjeshRPai committed:
* Fixes: Relative time not shown in minutes

zwarm committed:
Refactor: reduce the length of the data path

zwarm commented on WordPress/src/main/java/org/wordpress/android/util/DateTimeUtilsWrapper.kt:
```diff
@@ -41,4 +42,15 @@ class DateTimeUtilsWrapper @Inject constructor(
+    * Use getRelativeTimeSpanString function if the date is in the future.
+    * Time spans in future are formatted like "In 42 minutes | In 2 days".
+    * */
+    fun getRelativeTimeSpanString(date: Date): String {
```
❤️

zwarm reviewed (approved):
@AjeshRPai - All works as expected. Dark/light mode, dates + relative time (future and past). 👍 

Once CI finishes I am going to merge. @osullivanchris - any UI issues you may find will be addressed in a separate PR - we want to keep development moving along. Thanks for understanding.

osullivanchris commented:
@zwarm @AjeshRPai this looks great 👍  3 small details (which are fine for another PR as mentioned)

- The background for the label (draft/scheduled/published) seems to be using the color from the iOS design rather than the Android design. For Android I tried to use something existing, and took it from the Reader tags. 
- The text color of the secondary text (in 2 hours etc) isn't the same as the other cards. I think again it has grabbed a value from the iOS design. Can we use the same text color as the other cards (e.g. posts). 
- This one is per design, but think we should tweak it. The rows are optically off centre vertically. Could we move everything up by 2dp while keeping the row height? Might mean reducing top padding by 2dp, adding bottom padding by 2dp, or similar

AjeshRPai commented:
Hey @osullivanchris 

I will implement these changes in a separate PR by tomorrow. I have added the improvements you mentioned in this [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/18352). 

cc: @zwarm
